### PR TITLE
Allow local build definition spec for remote builds

### LINKED ISSUE #2262: Cant `build --remote` from a definition file
## Version of Singularity:

```
3.0.0-177.g945cf6e1
```

<br>

### The definition file:

```
Bootstrap: docker
From: ubuntu:18.04

%post
apt-get -y update
apt-get -y install curl
```

<br>

<br>

### Expected behavior:

```
$ singularity build --remote ubuntu.sif ubuntu.def
Getting image source signatures
[...]
Writing manifest to image destination
Storing signatures
INFO:    Creating SIF file...
INFO:    Build complete: /tmp/image-464963659
 26.68 MiB / 26.68 MiB [=======================================================================================] 100.00% 1.38 MiB/s 19s
$ 
```

<br>

### Actual behavior:

```
$ singularity build --remote ubuntu.sif ubuntu.def 
# nothing happened...
$ 
```

<br>

### Other information:

Builds just fine without a definition file.

```
$ singularity build --remote ubuntu.sif docker://ubuntu:16.04
[...]
Writing manifest to image destination
Storing signatures
INFO:    Creating SIF file...
INFO:    Build complete: /tmp/image-396692210
 35.33 MiB / 35.33 MiB [=======================================================================================] 100.00% 1.97 MiB/s 17s
$ 
```

### PR DESCRIPTION
**Pull request by mikegray**
**Description of the Pull Request (PR):**

This PR enables using the remote builder with local build definition files.

**This fixes or addresses the following GitHub issues:**

- Fixes #2262 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: src/cmd/singularity/cli/build_linux.go [hunk 1]
```diff
@@ -42,9 +41,9 @@ func run(cmd *cobra.Command, args []string) {
 			sylog.Fatalf("Unable to submit build job: %v", authWarning)
 		}
 
-		def, err := types.NewDefinitionFromURI(spec)
+		def, err := definitionFromSpec(spec)
 		if err != nil {
-			return
```
Comment (mikegray) [inline src/cmd/singularity/cli/build_linux.go:47]: This return is why you were getting silent failures - no debug output, or anything at all, just done.

File: src/cmd/singularity/cli/build_linux.go [hunk 2]
```diff
@@ -42,9 +41,9 @@ func run(cmd *cobra.Command, args []string) {
 			sylog.Fatalf("Unable to submit build job: %v", authWarning)
 		}
 
-		def, err := types.NewDefinitionFromURI(spec)
```
Comment (mikegray) [inline src/cmd/singularity/cli/build_linux.go:45]: In this "is remote" branch, only URI specs were being supported.

**Review by ikaneshiro — COMMENTED** [review]
Morning @mikegray! Take a look at `MakeDef()` in `/src/pkg/build/build.go`. I think this will cover the functionality that you want with `definitionFromSpec()`. Other than that this looks good to me

**Comment by mikegray** [comment]
@ikaneshiro Thanks for the review. I did see MakeDef() and used it as inspiration, but unfortunately I can't use it as-is because it pulls in `buildcfg` which needs code to be generated to fulfill the `LIBDIR` value. Anything that pulls in buildcfg will not work for macOS since we are not yet using mconfig for compilation for that platform.

**Review by ikaneshiro — APPROVED** [review]
Makes sense to me

**Review by tri-adam — APPROVED** [review]

**Pull request merged**